### PR TITLE
Basic documentation for Firmware Update feature

### PR DESCRIPTION
**Pull request by miha-plesko**
We describe basic usage of the fancy Firmware Update Feature.

### CODE REVIEW

File: redfish_guide/master.adoc [hunk 1]
```diff
@@ -1,5 +1,5 @@
 :imagesdir:    ../images
-:vernum:       0.2
+:vernum:       0.3
```
Comment (matejart) [inline redfish_guide/master.adoc]: No need to increment the version, we are still writing 0.2.

File: redfish_guide/topics/prerequisites.adoc [hunk 2]
```diff
@@ -15,6 +15,10 @@ The following preprequisites must be met to start using {redfish-provider-full}:
 * The target systems must be able to perform a network boot on at least one of
   their system's NICs (not the BMC network). This NIC must be able to reach the
   PXE server.
+* *Firmware Update feature*: One or more functional firmware registries must be configured or else Firmware
+  Update feature won't be available because there will be no firmware binary sources
+  available to install from. Furthermore, firmware binaries have to contain
+  precise list of supported target server models.
```
Comment (matejart) [inline redfish_guide/topics/prerequisites.adoc]: This is not a "Physical system provisioning additionally requires" item. I suggest adding a new paragraph with just "Firmware Update feature requires:" Then I would just itemize what needs to be available: a service that provides the firmware registry, a set of vendor-provided firmware update binary files.

Everything else is a process in ManageIQ that we should describe in this document. For the actual firmware registry, we can only prescribe what it does.

File: redfish_guide/topics/redfish-release-notes.adoc [hunk 3]
```diff
@@ -81,12 +82,16 @@ Operating System installations.
...
-physical systems. The Firmware Updates functionality consists of the following:
-
-* Firmware Registry contains the list of the images, binaries and executables
-  that are the payload in the firmware update process. The registry provides
-  the inventory of the available firmware updates by vendor, system and model.
-* Update Firmware of Physical Systems Lifecycle request, enabling requests for
-  performing the firmware updates on the selected servers.
\ No newline at end of file
+=== Physical Server Firmware Update
+{product-title} enables performing firmware update of the supported {redfish}
```
Comment (matejart) [inline redfish_guide/topics/redfish-release-notes.adoc]: `{product-title}` -> `{redfish-provider}`

File: redfish_guide/topics/redfish-release-notes.adoc [hunk 4]
```diff
@@ -81,12 +82,16 @@ Operating System installations.
...
+physical systems. The Firmware Update functionality consists of the following:
+
+* *Firmware Registry integration*. Firmware Registry is an external service that contains
+  a list of binaries that are the payload in the firmware update process. The registry
+  integration provides the inventory of the available firmware binaries tagged by
+  manufacturer and system model. Only compatible binaries are offered to install on a
+  specific server.
+* *Firmware Update request*. Firmware Update process is implemented in a form of
+  {product-title_short} request which can be initialized either from physical server details
+  page or list page. The latter allows one to trigger firmware update process on a number
```
Comment (matejart) [inline redfish_guide/topics/redfish-release-notes.adoc]: `trigger` -> `initiate`

File: redfish_guide/topics/using-redfish-physical-infra.adoc [hunk 5]
```diff
@@ -195,3 +195,65 @@ the selected image and customization template, has set the servers to boot once
...
+physical systems. The Firmware Update functionality consists of the following:
+
+* *Firmware Registry integration*. Firmware Registry is an external service that contains
+  a list of binaries that are the payload in the firmware update process. The registry
+  integration provides the inventory of the available firmware binaries tagged by
+  manufacturer and system model. Only compatible binaries are offered to install on a
+  specific server.
+* *Firmware Update request*. Firmware Update process is implemented in a form of
+  {product-title_short} request which can be initialized either from physical server details
+  page or list page. The latter allows one to trigger firmware update process on a number
```
Comment (matejart) [inline redfish_guide/topics/using-redfish-physical-infra.adoc]: `trigger` -> `initiate`

File: redfish_guide/topics/using-redfish-physical-infra.adoc [hunk 6]
```diff
@@ -195,3 +195,65 @@ the selected image and customization template, has set the servers to boot once
...
+  of servers in one go.
```
Comment (matejart) [inline redfish_guide/topics/using-redfish-physical-infra.adoc]: While I agree that reiterating a message from the release notes is useful, I'm not a fan of having copy&pasted text.

File: redfish_guide/topics/using-redfish-physical-infra.adoc [hunk 7]
```diff
@@ -195,3 +195,65 @@ the selected image and customization template, has set the servers to boot once
...
+
+NOTE: This guide assumes we have firmware registry already set up and registered with
+{product-title_short}. Moreover, this guide assumes registry contains at least one
+binary compatible with the physical server we are targeting for update.
```
Comment (matejart) [inline redfish_guide/topics/using-redfish-physical-infra.adoc:292]: This guide should provide instructions on registring a firmware registry with `{product-appliance}`.

File: redfish_guide/topics/using-redfish-physical-infra.adoc [hunk 8]
```diff
@@ -195,3 +195,65 @@ the selected image and customization template, has set the servers to boot once
...
+*Lifecycle* menu.
+
+image::firmware_update/center_menu.png[alt="Lifecycle menu for Firmware Update of Selected Physical Servers",550,align="center"]
+
+Click the *Lifecycle -> Update Firmware of Physical Servers*. This opens the
+dialog for filling in the request.
+
+image::firmware_update/popup.png[alt="Update Physical Server's Firmware dialog",350,align="center"]
+
+NOTE: Same dialog can be accessed from withing physical server's details page.
```
Comment (matejart) [inline redfish_guide/topics/using-redfish-physical-infra.adoc]: withing -> within

File: redfish_guide/topics/using-redfish-physical-infra.adoc [hunk 9]
```diff
@@ -195,3 +195,65 @@ the selected image and customization template, has set the servers to boot once
...
+
+The dialog informs us on the number of servers selected for this request. It
+presents the following fields that we need to fill:
```
Comment (matejart) [inline redfish_guide/topics/using-redfish-physical-infra.adoc]: Unless we plan to add other fields, I'd put this in singular: "It presents only one field:"

File: redfish_guide/topics/using-redfish-physical-infra.adoc [hunk 10]
```diff
@@ -195,3 +195,65 @@ the selected image and customization template, has set the servers to boot once
...
+
+* *Firmware Binary*: selection of the firmware binary to install on selected physical servers
+
+image::firmware_update/popup_filled.png[alt="Update Physical Server's Firmware dialog (filled)",350,align="center"]
+
+NOTE: Only compatible firmware binaries are available to choose from based on your current
+selection of physical servers. If you don't have any firmware binaries
+listed this means that either you don't have Firmware Registry set up correctly (visit
+*Compute -> Infrastructure -> Firmware Registry* to set it up) or that there are simply no
+compatible firmware binaries available for selected physical server.
```
Comment (matejart) [inline redfish_guide/topics/using-redfish-physical-infra.adoc]: for selected -> for the selected

The note reverses from "we" to "you". I suggest to unify to "we".

File: redfish_guide/topics/using-redfish-physical-infra.adoc [hunk 11]
```diff
@@ -195,3 +195,65 @@ the selected image and customization template, has set the servers to boot once
...
+
+
+Select desired firmware binary from the list and click *Apply Firmware*.
+
+image::firmware_update/request_created.png[alt="Firmware Update request created",350,align="center"]
```
Comment (matejart) [inline redfish_guide/topics/using-redfish-physical-infra.adoc]: Text too small. I suggest to change 350 to 1227 or, even better, crop to show just the panel with the pop-up.

Note, I find that a fairly good visibility of images is when I set the width limit to 0.75 times the image width in pixels.

File: redfish_guide/topics/using-redfish-physical-infra.adoc [hunk 12]
```diff
@@ -195,3 +195,65 @@ the selected image and customization template, has set the servers to boot once
...
+
+Firmware Update request gets executed without manual approval (it gets auto-approved). You can
+track its progress by visiting *Services -> Requests* and look for the request of type
+Physical Server Firmware Update.
+
+image::firmware_update/request_list_view.png[alt="Firmware Update request completed (list view)",500,align="center"]
```
Comment (matejart) [inline redfish_guide/topics/using-redfish-physical-infra.adoc]: This image is difficult to read too.

**Review by matejart — CHANGES_REQUESTED** [review]
Thank you for the documentation. Please see my comments on some of the improvement suggestions.

Please also provide a more detailed prerequisites section on the firmware updates registry capabilities, and describe the process of adding the firmware updates registry into `{product-name}`.

File: redfish_guide/topics/prerequisites.adoc (hunk 2 above)
Comment (miha-plesko) [reply redfish_guide/topics/prerequisites.adoc]: Yeah, the paragraph division of the prerequisites happened after I've opened the PR so it looks funny now when github shows it rebased already :) Fixed.

File: redfish_guide/topics/using-redfish-physical-infra.adoc (hunk 6 above)
Comment (miha-plesko) [reply redfish_guide/topics/using-redfish-physical-infra.adoc]: Yeah. Now that you pointed out, I actually found a neat solution to this: adoc supports "named chunks of text" which you can reuse later with `include::`. I've updated the code to be using that, please see.

File: redfish_guide/topics/using-redfish-physical-infra.adoc (hunk 7 above)
Comment (miha-plesko) [reply redfish_guide/topics/using-redfish-physical-infra.adoc:292]: Thank you for reading the TODO file out loud for me :) I think that's something that needs its own PR so I suggest we do it later as a followup

File: redfish_guide/topics/using-redfish-physical-infra.adoc (hunk 8 above)
Comment (miha-plesko) [reply redfish_guide/topics/using-redfish-physical-infra.adoc]: 😱

File: redfish_guide/topics/using-redfish-physical-infra.adoc (hunk 10 above)
Comment (miha-plesko) [reply redfish_guide/topics/using-redfish-physical-infra.adoc]: Good point, thanks

File: redfish_guide/topics/using-redfish-physical-infra.adoc (hunk 11 above)
Comment (miha-plesko) [reply redfish_guide/topics/using-redfish-physical-infra.adoc]: Set to 1227

File: redfish_guide/topics/using-redfish-physical-infra.adoc (hunk 12 above)
Comment (miha-plesko) [reply redfish_guide/topics/using-redfish-physical-infra.adoc]: FYI the size, the content and the readability of the image matches exactly what we have for "Provisioning request provisioned" in the previous chapter. But I agree, we can enlarge the image to 800 and it's much better readable right away.

**Review by miha-plesko — COMMENTED** [review]
Thanks for review @matejart , I've applied your suggestions and pushed. Please take a look.

File: redfish_guide/topics/using-redfish-physical-infra.adoc (hunk 12 above)
Comment (matejart) [reply redfish_guide/topics/using-redfish-physical-infra.adoc]: You're right, I need to make those bigger, too.

**Comment by matejart** [comment]
Looks good, but let's indeed follow up with the additional instructions.

**Pull request merged**